File: 2024-dgu-spring-camp/11/4703.py
# 동전 금고
# https://study.helloalgo.co.kr/study/1019/room/598/6561/source/828430/

# 높이가 x인 사다리를 가진 채 동전을 찾을 수 있는가?
from collections import deque

dirs = [
    (-1, 0),
    (0, -1),
    (1, 0),
    (0, 1),
]

N, M = map(int, input().split())
board = []
for _ in range(N):
    board.append(list(map(int, input().split())))
    
def isPossible(ladder):
    q = deque()
    visited = set()
    
    q.append((0, 0))
    visited.add((0, 0))
    while len(q) > 0:
        cx, cy = q.popleft()
        if cx == N-1 and cy == M-1:
            return True
        
        for dx, dy in dirs:
            nx, ny = cx+dx, cy+dy
            if 0<=nx<N and 0<=ny<M and board[cx][cy] + ladder >= board[nx][ny] and (nx, ny) not in visited:
                q.append((nx, ny))
                visited.add((nx, ny))
                
    return False
    
start, end = 0, 1_000_000_000
answer = -1
while start <= end:
    mid = (start + end)//2
    
    if isPossible(mid):
        answer = mid
        end = mid - 1
    else:
        start = mid + 1
        
print(answer)

File: 2024-dgu-spring-camp/11/4703.cpp
// 동전 금고
// https://study.helloalgo.co.kr/study/1019/room/598/6561

#include <iostream>
#include <vector>
#include <queue>
using namespace std;

vector<vector<int>> board;
int dx[] = {-1, 1, 0, 0};
int dy[] = {0, 0, -1, 1};

bool isPossible(int n, int m, int ladder)
{
    queue<pair<int, int>> q;
    bool visited[1001][1001];
    for (int x = 0; x < 1001; x++)
    {
        for (int y = 0; y < 1001; y++)
        {
            visited[x][y] = false;
        }
    }

    q.push({0, 0});
    visited[0][0] = true;
    while (!q.empty())
    {
        pair<int, int> cur = q.front();
        q.pop();
        int cx = cur.first, cy = cur.second;
        if (cx == n - 1 && cy == m - 1)
        {
            return true;
        }

        for (int di = 0; di < 4; di++)
        {
            int nx = cx + dx[di], ny = cy + dy[di];
            if (0 <= nx && nx < n && 0 <= ny && ny < m && board[cx][cy] + ladder >= board[nx][ny] && !visited[nx][ny])
            {
                q.push({nx, ny});
                visited[nx][ny] = true;
            }
        }
    }

    return false;
}

int main()
{
    int N, M, temp;
    cin >> N >> M;
    board.resize(N);
    for (int i = 0; i < N; i++)
    {
        board[i].resize(M);
        for (int j = 0; j < M; j++)
        {
            cin >> temp;
            board[i][j] = temp;
        }
    }

    int start = 0, end = 1000000000;
    int answer = -1;
    while (start <= end)
    {
        int mid = (start + end) / 2;

        if (isPossible(N, M, mid))
        {
            answer = mid;
            end = mid - 1;
        }
        else
        {
            start = mid + 1;
        }
    }

    cout << answer << '\n';

    return 0;
}
